python/arcticdb/version_store/_normalization.py

```diff
@@ -660,7 +698,15 @@ def denormalize(self, item, norm_meta):
         columns, denormed_columns, data = _denormalize_columns(item, norm_meta, idx_type, n_indexes)
 
         if not self._skip_df_consolidation:
+            columns_dtype = {} if data is None else {name: np_array.dtype for name, np_array in data.items()}
             df = DataFrame(data, index=index, columns=columns)
+
+            # Horribly setting the columns' dtype manually, since pandas might just convert the dtype of
+            # some (empty) columns to another one and since the `dtype` keyword for `pd.DataFrame` constructor
+            # does not accept a mapping such as `columns_dtype`...
+            for column_name, dtype in columns_dtype.items():
+                df[column_name] = df[column_name].astype(dtype, copy=False)
```

Comment:
An idea just occurred to me: can you try converting `columns` to a sequence of `pandas.Series` first and then construct the DF with `copy=False`?

Reply:
I tried this:

```diff
diff --git i/python/arcticdb/version_store/_normalization.py w/python/arcticdb/version_store/_normalization.py
index 280bbaf6..4fab43ba 100644
--- i/python/arcticdb/version_store/_normalization.py
+++ w/python/arcticdb/version_store/_normalization.py
@@ -698,14 +698,8 @@ class DataFrameNormalizer(_PandasNormalizer):
         columns, denormed_columns, data = _denormalize_columns(item, norm_meta, idx_type, n_indexes)
 
         if not self._skip_df_consolidation:
-            columns_dtype = {} if data is None else {name: np_array.dtype for name, np_array in data.items()}
-            df = DataFrame(data, index=index, columns=columns)
-
-            # Horribly setting the columns' dtype manually, since pandas might just convert the dtype of
-            # some (empty) columns to another one and since the `dtype` keyword for `pd.DataFrame` constructor
-            # does not accept a mapping such as `columns_dtype`...
-            for column_name, dtype in columns_dtype.items():
-                df[column_name] = df[column_name].astype(dtype, copy=False)
+            columns_mapping = {} if data is None else {name: pd.Series(np_array, index=index, dtype=np_array.dtype) for name, np_array in data.items()}
+            df = DataFrame(index=index, columns=columns_mapping, copy=False)
 
         else:
             if index is not None:

```

but this creates a `pd.DataFrame` full of NaNs whilst Series are sane. :shrug:

Reply:
I think that _skip_df_normalization is always true, and we want it to be, as the alternative is slow: https://github.com/man-group/ArcticDB/blob/002fa24473cbf222825c65fb6c08c6d8904b1669/python/arcticdb/version_store/library.py#L292

Reply:
Might the consolidation not actually be always skipped due to this definition?

https://github.com/man-group/ArcticDB/blob/6f5b9146f3e1f72ae40bbda7bbcca491c3a81a91/python/arcticdb/version_store/_normalization.py#L662

Reply:
Yes. I think we should remove this entirely at some later point, as we always want to skip it